### PR TITLE
[train] Storage: Call callbacks with new checkpoint

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Adjust checkpoint interface to accept new-style checkpoint objects instead.

In progress reporter output:

```
Training saved a checkpoint for iteration 4 at: (s3)bucket/foo/bar
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/experiment/trial.py [hunk 1]
```diff
@@ -1074,7 +1074,7 @@ def clear_checkpoint(self):
         self.temporary_state.restoring_from = None
         self.run_metadata.invalidate_cache()
 
-    def on_checkpoint(self, checkpoint: _TrackedCheckpoint):
+    def on_checkpoint(self, checkpoint: Union[_TrackedCheckpoint, _TrainingResult]):
```
Comment (justinvyu) [inline python/ray/tune/experiment/trial.py:1077]: Ahh I haven't been so on top of updating all of these 😄

**Review by justinvyu — COMMENTED** [review]
Thanks! One suggestion to fix the test.

File: python/ray/tune/experimental/output.py [hunk 2]
```diff
@@ -766,21 +767,29 @@ def on_checkpoint(
...
-        # don't think this is supposed to happen but just to be save.
+        # don't think this is supposed to happen but just to be safe.
         saved_iter = "?"
         if trial.last_result and TRAINING_ITERATION in trial.last_result:
             saved_iter = trial.last_result[TRAINING_ITERATION]
 
         self._start_block(f"trial_{trial}_result_{saved_iter}")
+
+        if isinstance(checkpoint, Checkpoint):
+            loc = f"({checkpoint.filesystem.type_name})"
```
Comment (ericl) [inline python/ray/tune/experimental/output.py:783]: Shouldn't this also include the path?

File: python/ray/tune/experimental/output.py [hunk 3]
```diff
@@ -766,21 +767,29 @@ def on_checkpoint(
...
-        # don't think this is supposed to happen but just to be save.
+        # don't think this is supposed to happen but just to be safe.
         saved_iter = "?"
         if trial.last_result and TRAINING_ITERATION in trial.last_result:
             saved_iter = trial.last_result[TRAINING_ITERATION]
 
         self._start_block(f"trial_{trial}_result_{saved_iter}")
+
+        if isinstance(checkpoint, Checkpoint):
+            loc = f"({checkpoint.filesystem.type_name}){checkpoint.path}"
```
Comment (krfricke) [inline python/ray/tune/experimental/output.py:783]: We currently have no way to restore the actual location URI, so this is the next best thing.

File: python/ray/tune/execution/tune_controller.py [hunk 4]
```diff
@@ -1981,6 +1981,7 @@ def _process_trial_save(
 
         try:
             if _use_storage_context() and isinstance(checkpoint_value, _TrainingResult):
+                assert isinstance(checkpoint_value, _TrainingResult), checkpoint_value
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:1984]: nit: can remove the part in the if condition

File: python/ray/tune/experimental/output.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/tune/experimental/output.py:783]: add an example of what this looks like in the pr description?

**Review by justinvyu — APPROVED** [review]
Thanks, lgtm. Only left some nits.

**Comment by ericl** [comment]
Good to merge, but just in case you want to address the nits.

**Pull request merged**